timemachine/cpp/src/barostat.cu

```diff
@@ -140,31 +142,31 @@ void MonteCarloBarostat::inplace_move(
         return;
     }
 
-    curandErrchk(curandSetStream(cr_rng_, stream));
-    // Generate scaling and metropolis conditions in one pass
-    curandErrchk(curandGenerateUniformDouble(cr_rng_, d_rand_, 2));
+    // Get offset into the d_rand_ array
+    int random_offset = (((step_ / interval_) * 2) - 2) % (RANDOM_BATCH_SIZE * 2);
 
-    runner_.execute_potentials(bps_, N_, d_x, d_box, nullptr, nullptr, d_u_buffer_, stream);
-    accumulate_energy(bps_.size(), d_u_buffer_, d_init_u_, stream);
-
-    k_setup_barostat_move<<<1, 1, 0, stream>>>(d_rand_, d_box, d_volume_delta_, d_volume_scale_, d_length_scale_);
-    gpuErrchk(cudaPeekAtLastError());
+    // Generate the scaling and metropolis conditions in batches then offset on each move
+    if (random_offset == 0) {
+        curandErrchk(curandSetStream(cr_rng_, stream));
+        curandErrchk(curandGenerateUniformDouble(cr_rng_, d_rand_, RANDOM_BATCH_SIZE * 2));
```

Comment:
should this still generate `Double` if we're switching to float for everything else?

Reply:
Templated as we discussed offline

# Performance differences between float and double

## Float
```
0.1          752,387         80    9,404.8    9,952.0     6,560    11,296      1,288.1  void k_decide_move<float>(int, int, double, double, const T1 *, T1 *, T1 *, const __int128 *, const…
0.1          682,754         80    8,534.4    8,513.0     8,095     9,088        191.0  void k_find_group_centroids<float>(int, const double *, const int *, const int *, unsigned long lon…
0.0          501,925         80    6,274.1    6,240.0     6,048     6,784        153.1  void k_rescale_positions<float>(int, double *, const T1 *, const double *, double *, const int *, c…
0.0          210,304         80    2,628.8    2,593.0     2,496     3,008         95.1  void k_setup_barostat_move<float>(const T1 *, double *, T1 *, T1 *, T1 *)
0.0            2,431          1    2,431.0    2,431.0     2,431     2,431          0.0  void gen_sequenced<curandStateXORWOW, float, int, &curand_uniform_noargs<curandStateXORWOW>, rng_co…
```
## Double
```
0.1        1,214,819         80   15,185.2   15,776.0    12,096    17,088      1,512.3  void k_decide_move<double>(int, int, double, double, const T1 *, T1 *, T1 *, const __int128 *, cons…
0.1          870,049         80   10,875.6   10,783.5    10,592    11,744        265.3  void k_rescale_positions<double>(int, double *, const T1 *, const double *, double *, const int *, …
0.1          667,139         80    8,339.2    8,320.0     8,064     8,896        206.4  void k_find_group_centroids<double>(int, const double *, const int *, const int *, unsigned long lo…
0.0          249,412         80    3,117.7    3,104.0     3,008     3,520         95.0  void k_setup_barostat_move<double>(const T1 *, double *, T1 *, T1 *, T1                                                    
0.0            2,528          1    2,528.0    2,528.0     2,528     2,528          0.0  void gen_sequenced<curandStateXORWOW, double, int, &curand_uniform_double_noargs<curandStateXORWOW>…
```

# Correctness
The question of whether computing in float vs double produces more correct values, I inspected the `test_molecular_ideal_gas` with the two different types. Not an exhaustive look, but while they do have some difference it doesn't seem clear that there is any real difference

## 10k steps
The test currently runs 10k steps

### float
Ideal Gas Float: [ 369.52854306  739.05708613 1231.74678328] Reference [ 367.807291, 735.614582, 1226.024304]
Relative error: 0.005

### double
Ideal Gas Double: [ 367.10525596,  734.21051192, 1223.68418653] Reference [ 367.807291, 735.614582, 1226.024304]
Relative error: 0.001

## 20k steps
Want to see if doubling the test results in a greater difference between the two, to see if it is different due to potential difference due to randomness or some shift

### float
Ideal Gas Float: [ 367.463538,  734.927076, 1224.878918] Reference [ 367.807291,  735.614582, 1226.024304]
Relative error: 0.0009

### double
Ideal Gas Double: [ 367.10525596,  734.21051192, 1223.68418653] [ 367.807291, 735.614582, 1226.024304]
Relative error: 0.001

Ideal Gas Double: [ 367.10525596,  734.21051192, 1223.68418653] [ 367.807291, 735.614582, 1226.024304]
Relative error: 0.001
Ideal Gas Double: [ 367.117415,  734.234831, 1223.724718] Reference [ 367.807291,  735.614582, 1226.024304]
Relative error: 0.001